src/backend.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright (C) Nitrokey GmbH
+// SPDX-License-Identifier: Apache-2.0 or MIT
+
+mod data;
+
+use trussed::{
+    backend::Backend,
+    error::Result,
+    platform::Platform,
+    serde_extensions::ExtensionImpl,
+    service::ServiceResources,
+    store::filestore::Filestore,
+    types::{CoreContext, Location, PathBuf},
+};
+
+use crate::{
+    extension::{reply, AuthExtension, AuthReply, AuthRequest},
+    PIN_PATH,
+};
+use data::PinData;
+
+pub struct AuthBackend {
+    location: Location,
+}
+
+impl AuthBackend {
+    pub fn new(location: Location) -> Self {
+        Self { location }
+    }
+}
+
+impl<P: Platform> Backend<P> for AuthBackend {
+    type Context = ();
+}
+
+impl<P: Platform> ExtensionImpl<AuthExtension, P> for AuthBackend {
+    fn extension_request(
+        &mut self,
+        core_ctx: &mut CoreContext,
+        _ctx: &mut (),
+        request: &AuthRequest,
+        resources: &mut ServiceResources<P>,
+    ) -> Result<AuthReply> {
+        let fs = &mut resources.filestore(core_ctx);
+        match request {
+            AuthRequest::HasPin(request) => {
+                let has_pin = fs.exists(&request.id.path(), self.location);
+                Ok(reply::HasPin { has_pin }.into())
+            }
+            AuthRequest::CheckPin(request) => {
+                // TODO: return error or bool?
```

Comment:
* the `Result` should be `bool`
 * an `Err` is only returned, if the checking operation failed

Reply:
What if the pin is blocked or not set?

Reply:
I’ve changed it to return `false` if the PIN is blocked or wrong.  Still not sure what to do if the PIN is not set.

Reply:
so, should we go for?
* the `Ok` `Result` can only be `true`
* any other case is an `Err` i.e., `BadPin`, `PinNotSet`, `PinBlocked` ?

Reply:
I don’t think that makes sense.  Either we have a `Result<()>`, where `Ok(())` means success (initial implementation).  Or we have a `Result<bool>` where `Ok(true)` means success, `Ok(false)` means failure and `Err(_)` means error (updated implementation, with the bool being wrapped in the reply type).

I think the current approach is okay, I’m just not sure if a missing PIN should be considered a failure or an error.

Reply:
Given yersterday's discussion to include the per-pin key derivation here, don't we want to include a `KeyId` pointing to a volatile 32 byte symmetric key after a successful validation?

Reply:
I think this should be done in a separate PR.  Also, I currently prefer doing that in a separate syscall to avoid unnecessary work for applications that don’t need the PIN-specific key.

Reply:
I’ve thought about the missing PIN issue some more.  I think having it as an error is the better option.  Calling `check_pin` on an unset PIN most likely is a programming error.  Either the PIN should always be set, or the application should explicitly handle the case of a missing PIN.